File: problemskeleton_Problem32/submissions/accepted/accepted.java
// Time Complexity: O(n)
// Space Complexity: O(n)
package COMP321.problemskeleton_Problem32.submissions.accepted;

import java.util.*;
import java.io.*;

public class accepted {
    static class Line {
        long m, c;
        
        Line(long m, long c) {
            this.m = m;
            this.c = c;
        }
        
        long eval(long x) {
            return m * x + c;
        }
        
        double intersectX(Line other) {
            if (m == other.m) {
                return c >= other.c ? Double.POSITIVE_INFINITY : Double.NEGATIVE_INFINITY;
            }
            return (double)(other.c - c) / (m - other.m);
        }
    }
    
    static boolean isBad(Line l1, Line l2, Line l3) {
        return l1.intersectX(l3) <= l1.intersectX(l2);
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int n = Integer.parseInt(br.readLine());
        
        long[] steps = new long[n];
        long[] complexities = new long[n];
        
        for (int i = 0; i < n; i++) {
            String[] parts = br.readLine().split(" ");
            steps[i] = Long.parseLong(parts[0]);
            complexities[i] = Long.parseLong(parts[1]);
        }
        
        long[] dp = new long[n];
        Arrays.fill(dp, Long.MAX_VALUE);
        dp[0] = 0;
        
        ArrayDeque<Line> hull = new ArrayDeque<>();
        
        hull.addLast(new Line(complexities[0], dp[0]));
        
        for (int i = 1; i < n; i++) {
            while (hull.size() >= 2) {
                Line[] arr = hull.toArray(new Line[0]);
                if (arr[0].eval(steps[i]) >= arr[1].eval(steps[i])) {
                    hull.pollFirst();
                } else {
                    break;
                }
            }
            
            if (!hull.isEmpty()) {
                dp[i] = hull.peekFirst().eval(steps[i]);
            }
            
            Line newLine = new Line(complexities[i], dp[i]);
            while (hull.size() >= 2) {
                Line last = hull.pollLast();
                Line secondLast = hull.peekLast();
                if (isBad(secondLast, last, newLine)) {
                    // Keep polling
                } else {
                    hull.addLast(last);
                    break;
                }
            }
            hull.addLast(newLine);
        }
        
        System.out.println(dp[n - 1]);
    }
}

File: problemskeleton_problem32/submissions/accepted/accepted.cpp
// Time Complexity: O(n)
// Space Complexity: O(n)
#include <iostream>
#include <vector>
#include <deque>
#include <climits>
using namespace std;

struct Line {
    long long m, c;
    
    long long eval(long long x) const {
        return m * x + c;
    }
    
    double intersect_x(const Line& other) const {
        if (m == other.m) {
            return c >= other.c ? 1e18 : -1e18;
        }
        return (double)(other.c - c) / (m - other.m);
    }
};

bool is_bad(const Line& l1, const Line& l2, const Line& l3) {
    return l1.intersect_x(l3) <= l1.intersect_x(l2);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n;
    cin >> n;
    
    vector<long long> steps(n), complexities(n);
    for (int i = 0; i < n; i++) {
        cin >> steps[i] >> complexities[i];
    }
    
    vector<long long> dp(n, LLONG_MAX);
    dp[0] = 0;
    
    deque<Line> hull;
    
    auto add_line = [&](Line line) {
        while (hull.size() >= 2 && is_bad(hull[hull.size()-2], hull[hull.size()-1], line)) {
            hull.pop_back();
        }
        hull.push_back(line);
    };
    
    auto query = [&](long long x) -> long long {
        while (hull.size() >= 2 && hull[0].eval(x) >= hull[1].eval(x)) {
            hull.pop_front();
        }
        return hull.empty() ? LLONG_MAX : hull[0].eval(x);
    };
    
    add_line({complexities[0], dp[0]});
    
    for (int i = 1; i < n; i++) {
        dp[i] = query(steps[i]);
        add_line({complexities[i], dp[i]});
    }
    
    cout << dp[n - 1] << endl;
    
    return 0;
}
